pkg/tracing/tracing.go

```diff
@@ -0,0 +1,144 @@
+
+		if caCert != "" {
+			data, err := os.ReadFile(caCert)
+			if err != nil {
+				klog.ErrorS(err, "failed to extract ca certificate required to generate the transport credentials")
+				return nil, err
+			}
+			pool := x509.NewCertPool()
+			if !pool.AppendCertsFromPEM(data) {
+				klog.Error("failed to create cert pool using the ca certificate provided for generating the transport credentials")
+				return nil, err
+			}
+			klog.Info("Enabling trace GRPC client in secure TLS mode")
+			opts = append(opts, otlptracegrpc.WithTLSCredentials(credentials.NewClientTLSFromCert(pool, "")))
+		} else {
+			klog.Info("Enabling trace GRPC client in insecure mode")
+			opts = append(opts, otlptracegrpc.WithInsecure())
+		}
+
+		client := otlptracegrpc.NewClient(opts...)
+
+		exporter, err := otlptrace.New(ctx, client)
+		if err != nil {
+			klog.ErrorS(err, "failed to create an instance of the trace exporter")
+			return nil, err
+		}
+		if name == "" {
+			klog.V(5).InfoS("no name provided, using default service name for tracing", "name", DefaultServiceName)
+			name = DefaultServiceName
+		}
+		resourceOpts := []sdkresource.Option{sdkresource.WithAttributes(semconv.ServiceNameKey.String(name)), sdkresource.WithSchemaURL(semconv.SchemaURL)}
+		if namespace != "" {
+			klog.V(5).InfoS("no namespace provided, using default namespace")
+			resourceOpts = append(resourceOpts, sdkresource.WithAttributes(semconv.ServiceNamespaceKey.String(namespace)))
+		}
+		resource, err := sdkresource.New(ctx, resourceOpts...)
+		if err != nil {
+			klog.ErrorS(err, "failed to create traceable resource")
+			return nil, err
+		}
+
+		spanProcessor := sdktrace.NewBatchSpanProcessor(exporter)
+		provider = sdktrace.NewTracerProvider(
+			sdktrace.WithSampler(sdktrace.ParentBased(sdktrace.TraceIDRatioBased(sampleRate))),
+			sdktrace.WithSpanProcessor(spanProcessor),
+			sdktrace.WithResource(resource),
+		)
+	} else {
+		klog.Info("Did not find a trace collector endpoint defined. Switching to NoopTraceProvider")
+		provider = trace.NewNoopTracerProvider()
+	}
+
+	otel.SetTextMapPropagator(propagation.TraceContext{})
+	otel.SetTracerProvider(provider)
+	otel.SetErrorHandler(otel.ErrorHandlerFunc(func(err error) {
+		klog.ErrorS(err, "got error from opentelemetry")
+	}))
+	klog.Info("Successfully setup trace provider")
+	return
+}
+
+// StartSpan starts a new span with the given operation name and attributes.
+func StartSpan(ctx context.Context, tracerName, operationName string) (context.Context, trace.Span, func()) {
+	return StartSpanWithOptions(ctx, tracerName, []trace.TracerOption{}, operationName, []attribute.KeyValue{})
+}
+
+func StartSpanWithOptions(ctx context.Context, tracerName string, traceOpts []trace.TracerOption, operationName string, attributes []attribute.KeyValue) (context.Context, trace.Span, func()) {
+	spanCtx, span := otel.Tracer(tracerName, traceOpts...).Start(ctx, operationName, trace.WithAttributes(attributes...))
```

Comment:
It looks like you are passing a different `tracerName` for each type of operation (ie, `descheduler_parent`, `balance`, `evict`) right? We can just use a single tracer for all of descheduler. Attributes like that should come from the spans themselves

We could actually store one tracer at startup and pass that variable through everywhere, maybe in the framework handle. By calling `otel.Tracer()` each time, you are making a new allocation (that's more a performance improvement though)

Reply:
> maybe in the framework handle.

Do you mean one plugin will get one Tracer with the name set for it ? Or for the entire runtime, there will be one tracer and every plugin will just use the same tracer ?

Reply:
one tracer for the entire runtime

Reply:
cc @harshanarayana